### PR TITLE
[SPARK-35543][CORE] Fix memory leak in BlockManagerMasterEndpoint removeRdd

### PR DESCRIPTION
**Pull request by attilapiros**
### What changes were proposed in this pull request?

In `BlockManagerMasterEndpoint` for the disk persisted RDDs (when `spark.shuffle.service.fetch.rdd.enable` is enabled) we are keeping track the block status entries by external shuffle service instances (so on YARN we are basically keeping them by nodes). This is the `blockStatusByShuffleService` member val. And when all the RDD blocks are removed for one external shuffle service instance then the key and the empty map can be removed from `blockStatusByShuffleService`.

### Why are the changes needed?

It is a small leak and I was asked to take care of it in https://github.com/apache/spark/pull/32114#discussion_r640270377.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Manually by adding a temporary log line to check `blockStatusByShuffleService` value before and after the `removeRdd` and run the `SPARK-25888: using external shuffle service fetching disk persisted blocks` test in `ExternalShuffleServiceSuite`.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:652]: After this PR `blockStatusByShuffleService.get(bmId)` can be `None` and even when it has some value for the key then even `m.get(blockId)` can be `null` as m is a Java `HashMap`.

**Comment by attilapiros** [comment]
I can see two unrelated errors:

1.
```
[info] - Non-consecutive stage failures don't trigger abort *** FAILED *** (635 milliseconds)
[info]   stageAttemptOpt.isDefined was false (DAGSchedulerSuite.scala:978)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
[info]   at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
[info]   at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
[info]   at org.apache.spark.scheduler.DAGSchedulerSuite.completeShuffleMapStageSuccessfully(DAGSchedulerSuite.scala:978)
[info]   at org.apache.spark.scheduler.DAGSchedulerSuite.$anonfun$new$59(DAGSchedulerSuite.scala:1211)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:190)
[info]   at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:190)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:188)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:200)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:200)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:182)
[info]   at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:62)
[info]   at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:234)
[info]   at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:227)
[info]   at org.apache.spark.SparkFunSuite.runTest(SparkFunSuite.scala:62)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:233)
[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
[info]   at scala.collection.immutable.List.foreach(List.scala:431)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:233)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:232)
[info]   at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1563)
[info]   at org.scalatest.Suite.run(Suite.scala:1112)
[info]   at org.scalatest.Suite.run$(Suite.scala:1094)
[info]   at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1563)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:237)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:237)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:236)
[info]   at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:62)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:62)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:318)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:513)
[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:413)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]   at java.lang.Thread.run(Thread.java:748)
```

2.
```
Moving to python/docs directory and building sphinx.
Running Sphinx v3.0.4

Extension error:
Could not import extension sphinx_plotly_directive (exception: No module named 'sphinx_plotly_directive')
make: *** [Makefile:20: html] Error 2
                    ------------------------------------------------
      Jekyll 4.2.0   Please append `--trace` to the `build` command 
                     for any additional information or backtrace. 
                    ------------------------------------------------
/__w/spark/spark/docs/_plugins/copy_api_dirs.rb:130:in `<top (required)>': Python doc generation failed (RuntimeError)
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.0/lib/jekyll/external.rb:60:in `require'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.0/lib/jekyll/external.rb:60:in `block in require_with_graceful_fail'
```

**Comment by attilapiros** [comment]
I think I just used an outdated master branch in my local clone. 
So I rebased my changed on the top of fresh master and reverted those Sphinx related commits of mine.

**Comment by attilapiros** [comment]
I am still not sure my github action fix is really needed so I will do another test when this one finishes...

**Comment by attilapiros** [comment]
cc @Ngone51 @mridulm

**Comment by attilapiros** [comment]
I just realized existing tests does not covers this so please let me run at least a manual test before the merge.

**Comment by attilapiros** [comment]
Actually the `"SPARK-25888: using external shuffle service fetching disk persisted blocks` test in `ExternalShuffleServiceSuite`  covers this case.

**Comment by mridulm** [comment]
Do we need to also handle `updateBlockInfo` @attilapiros ?

**Comment by attilapiros** [comment]
@mridulm the `updateBlockInfo` does not use  `blockStatusByShuffleService`. Isn't it?

It uses `blockLocation` which is another data structure where for each block ID we are storing the block manager IDs where the block is available.

**Comment by attilapiros** [comment]
You probably meant he `updateBlockInfo` of the `BlockManagerInfo`. Let me check that!

**Comment by attilapiros** [comment]
That one is also negative. It uses an ` externalShuffleServiceBlockStatus: Option[JHashMap[BlockId, BlockStatus]]` where entries are removed by the key:
https://github.com/apache/spark/blob/7be0ce45a95b4f2449ac370baceb0b9f461e7199/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L890

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:281]: nit: can we add a comment about why we remove it here? (e.g the memory leak) because just reading the code through it might not be super obvious.

**Review by holdenk — APPROVED** [review]
LGTM one minor optional suggestion re: comment

**Comment by Ngone51** [comment]
https://github.com/apache/spark/blob/b8740a1d1ef6f7e08873a37c7ff1b4e3980679f0/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L877-L881
^^^ If `BlockManagerInfo.externalShuffleServiceBlockStatus` only handles RDD blocks, shall we use `if (blockId.isRDD && blockStatus.diskSize > 0)` instead to avoid mistakenly leak for other block types?

**Comment by attilapiros** [comment]
@Ngone51 currently I cannot see a leak here but I am open to change this if it is proven to be exists.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by Ngone51** [comment]
> @Ngone51 currently I cannot see a leak here but I am open to change this if it is proven to be exists.

**Comment by attilapiros** [comment]
> However, looking at removeBlock, it shows it only cleans rdd blocks indeed. So it makes me think whether we missed other block types. Fortunately, we only have broadcast block and rdd block here so it won't leak anything. 

@Ngone51 I am not so sure about that.

1.)
> However, looking at removeBlock, it shows it only cleans rdd blocks indeed.

Why do you sure about that? Just looking only these lines?

https://github.com/apache/spark/blob/6de20f10ca60c778db25ba3d5975f85e5c23aedb/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L906-L914

2.
> Fortunately, we only have broadcast block and rdd block here so it won't leak anything. 

Do we? 

https://github.com/apache/spark/blob/6de20f10ca60c778db25ba3d5975f85e5c23aedb/core/src/main/scala/org/apache/spark/storage/BlockId.scala#L96

Related to both points: 
https://github.com/apache/spark/blob/6de20f10ca60c778db25ba3d5975f85e5c23aedb/core/src/main/scala/org/apache/spark/executor/Executor.scala#L617
https://github.com/apache/spark/blob/6de20f10ca60c778db25ba3d5975f85e5c23aedb/core/src/main/scala/org/apache/spark/scheduler/TaskResultGetter.scala#L80


I am sorry but now I have a limited time for this (starting tomorrow I will be without my laptop for the next 10 days) and within the scope "Fix memory leak in BlockManagerMasterEndpoint removeRdd" I intend to fix the existing / proved memory leaks and avoid introducing any new errors or performance degradation by being not careful enough. Moreover I do not want to leave any potential error for the others (unattended by me). With my phone I cannot fix it (although a vim probably would run just fine on it ;) but having no keyboard no sbt..) so it would disturb me a lot and it could ruin my vacation what I do not want risk.

**Comment by attilapiros** [comment]
K8s failures are unrelated: https://github.com/apache/spark/pull/32790#issuecomment-855590776

**Comment by Ngone51** [comment]
> Why are you sure about that? Just looking only these lines?

**Comment by Ngone51** [comment]
Ok, l see there's another type of block - `TaskResultBlockId`. So it looks good!!

**Review by Ngone51 — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you, @attilapiros and all. Could you backport this, @attilapiros ?

**Comment by attilapiros** [comment]
@holdenk @dongjoon-hyun @Ngone51 
Thanks for the reviews!

**Comment by attilapiros** [comment]
@dongjoon-hyun  I am sorry I have to run now and I thought during the merge this minor leak won't worth it. 
If you think it is needed may I ask a favour to backport it when you have time?

**Comment by mridulm** [comment]
@attilapiros:

> That one is also negative. It uses an externalShuffleServiceBlockStatus: Option[JHashMap[BlockId, BlockStatus]] where entries are removed by the key:

I was referring to this, yes - essentially when all the block id's are removed - we end up with an entry in `blockManagerInfo` Map (which is referring to this hash map as the value) which does not have any entries (empty map).

Essentially, this is similar as `removeRdd`.
In `BlockManagerMasterEndpoint.updateBlockInfo`, after `blockManagerInfo(blockManagerId).updateBlockInfo` we can end up with an empty map as value.

**Comment by attilapiros** [comment]
@mridulm sorry I cannot focus on this right now. 
I create a reminder in my calendar to come back to this when I am back from my Holiday and about the backport too (cc @dongjoon-hyun).

**Comment by mridulm** [comment]
@attilapiros No hurries, enjoy your break :-)

**Comment by dongjoon-hyun** [comment]
No problem. Thank you, @attilapiros ! Enjoy your holiday.

**Comment by attilapiros** [comment]
I think I made a mistake here and we should revert this change.

But let me share my thoughts and discuss the problem.

So we have `blockStatusByShuffleService` for keeping tracks of the block statuses for external shuffle services:

https://github.com/apache/spark/blob/6de20f10ca60c778db25ba3d5975f85e5c23aedb/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L64-L66

The `JHashMap[BlockId, BlockStatus]` part is shared with the block infos when a new block manager is registered:

https://github.com/apache/spark/blob/6de20f10ca60c778db25ba3d5975f85e5c23aedb/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L570-L580

When the feature is on and this is the first block manager on a node a new `JHashMap[BlockId, BlockStatus]` is created. 

Although `removeRdd` is not routed to the  blockmanger infos but update of the block status is done in the method:
https://github.com/apache/spark/blob/6de20f10ca60c778db25ba3d5975f85e5c23aedb/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L823-L827

So what happens when we add an disk persisted RDD and remove it latter then add a new one then lost the Block Manager. In that case `getLocationsAndStatus()` gives back no block infos even this could be given back via the shuffle service. So this fails:

https://github.com/apache/spark/blob/6de20f10ca60c778db25ba3d5975f85e5c23aedb/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L672-L675

There is no lifecycle management for nodes but for Block Managers (executors) so we cannot know when we get a new BM using the earlier node.

We could remove the map and when a new block is persisted recreate the map then re-register it to every block infos related to BMs running on the node but does this worth the work? In addition to check every time whether the map exists and the feature is on when we would like to add an entry? This has run time cost and I do not think it worth it compared to the memory footprint of this "leak" which is proportional to the number of nodes.

@mridulm  @Ngone51  WDYT?

**Comment by dongjoon-hyun** [comment]
It's fine to revert this because this is not released yet. Can we have a test case, @attilapiros ?

BTW, cc @gengliangwang since he is the release manager for Apache Spark 3.2.0.

**Comment by attilapiros** [comment]
@dongjoon-hyun it is hard to test this as `BlockManagerMasterEndpoint` is not tested directly. So first I would revert this then we can think about how to cover these cases with tests. 
But before the revert I am interested what the others thinks about this.

**Comment by mridulm** [comment]
My main concern was around long running applications which cycle through a lot of executors (over time - not concurrently).
Will this not cause issues there eventually ?

**Comment by attilapiros** [comment]
@mridulm could we estimate the number of nodes for those long running applications?
This is proportional to the number of nodes/hosts not with the number of executors.

**Comment by mridulm** [comment]
(Ignoring ephemenral nodes, dockerized deployments, etc - and using more real world scenarios)

For a 5k cluster, the map can have upto 5k entries - with the values maps going of reasonable size (even if empty).
For example, assuming 1k blocks per node on average - we end up with the map using approximately: ~ 8192 * 2048 * (~8 bytes) == ~128MB : this is the driver cost with zero blocks being used.

**Comment by attilapiros** [comment]
> (Ignoring ephemenral nodes, dockerized deployments, etc - and using more real world scenarios)

The feature ([Serve local disk persisted blocks by the external service](https://github.com/apache/spark/pull/24499)) only works when external shuffle service is available.

> For example, assuming 1k blocks per node on average - we end up with the map using approximately: ~ 5000 * 2048 * (~8 bytes) == ~80MB : this is the driver cost with zero blocks being used.

I do not think the number of blocks matters at all. So `blockStatusByShuffleService` is the following data structure:
https://github.com/apache/spark/blob/6de20f10ca60c778db25ba3d5975f85e5c23aedb/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L64-L66

It maps a special `BlockManagerId` instances which are created for the external shuffle service (so for the nodes) to a map which maps `BlockId` instances to a BlockStatus which is a simple case class:
https://github.com/apache/spark/blob/37ef7bb98cdb1a8eefa06677f119a4d97e242097/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L788-L790

So the number is ~ 5000 * 2 * (~8 bytes) = 80KB = 0.08MB

**Comment by mridulm** [comment]
I am referring to the size of the Map's themselves @attilapiros.
Essentially, sum of sizes of each value `JHashMap`.

**Comment by attilapiros** [comment]
Your example assumes all the blocks was persisted into disk. In that case when during the run of the application when persist is called again those area will be reused. 

Even though I have an idea: introducing a new class which wraps the `JHashMap` and when the HashMap is empty we can create remove the reference to it and when it is needed again we can recreate the HashMap.
Let me look into that.

**Comment by Ngone51** [comment]
@mridulm Interesting example! I also investigate a bit on it. Looks like all the elements of the underlying array has been nulled out but remain the array refrence unchanged. So its size doesn't change. But the memory usage shouldn't be the same as before but it's also not empty since `null` still takes a bit memory.

```java
public void clear() {
    Node<K,V>[] tab;
    modCount++;
    if ((tab = table) != null && size > 0) {
        size = 0;
        for (int i = 0; i < tab.length; ++i)
            tab[i] = null;
    }
}
```

**Comment by mridulm** [comment]
@Ngone51 Yes, the table contents are null'ed, but the table itself remains and occupies memory.
I used to use gnu trove4j (for example) to mitigate these sorts of issues (and to use primitive collection when relevant ofcourse !).

**Pull request closed without merging**
